codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java

```diff
@@ -0,0 +1,182 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.ChannelPromiseNotifier;
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.OneTimeTask;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Compresses a {@link ByteBuf} using one of compression algorithms.
+ */
+public abstract class CompressionEncoder extends MessageToByteEncoder<ByteBuf> {
+    /**
+     * Default delay time for close. See method {@link #closeDelay(int)}.
+     */
+    public static final int DEFAULT_CLOSE_DELAY = 10;
+
+    private volatile int closeDelay;
+
+    private final CompressionFormat format;
+
+    /**
+     * Used to interact with its {@link ChannelPipeline} and other handlers.
+     */
+    private volatile ChannelHandlerContext ctx;
+
+    /**
+     * Default constructor for {@link CompressionEncoder} which will try to use a direct
+     * {@link ByteBuf} as a target for the encoded messages.
+     */
+    protected CompressionEncoder(CompressionFormat format) {
+        this(format, true);
+    }
+
+    /**
+     * Creates an instance with specified preference of a target {@link ByteBuf} for the encoded messages.
+     * It uses if compression algorithm needs a byte array as a target.
+     *
+     * @param preferDirect {@code true} if a direct {@link ByteBuf} should be tried to be used as a target
+     *                     for the encoded messages. If {@code false} is used it will allocate a heap
+     *                     {@link ByteBuf}, which is backed by an byte array.
+     */
+    protected CompressionEncoder(CompressionFormat format, boolean preferDirect) {
+        super(preferDirect);
+        ObjectUtil.checkNotNull(format, "format");
+        this.format = format;
+        this.closeDelay = DEFAULT_CLOSE_DELAY;
+    }
+
+    /**
+     * Returns a format of current compression algorithm.
+     */
+    public final CompressionFormat format() {
+        return format;
+    }
+
+    /**
+     * Set amount of time to be delayed before the ensuring that the channel is closed
+     * after the call of {@link #close(ChannelHandlerContext, ChannelPromise)}.
+     *
+     * The default value is {@link #DEFAULT_CLOSE_DELAY}.
+     *
+     * @param delay time in seconds.
```

Comment:
- Could we make this millis?
- Could we rename this property to something better? How about `lastWriteTimeoutMillis`?